File: src/data/courses/dsa/queues/deque/implement-deque/code/python/implement_deque.py
class Deque:
    def __init__(self, k):
        self.arr = [0] * k
        self.capacity = k
        self.front = self.rear = -1
        self.size = 0
    
    def insert_front(self, x):
        if self.size == self.capacity:
            return False
        if self.front == -1:
            self.front = self.rear = 0
        else:
            self.front = (self.front - 1 + self.capacity) % self.capacity
        self.arr[self.front] = x
        self.size += 1
        return True
    
    def insert_rear(self, x):
        if self.size == self.capacity:
            return False
        if self.rear == -1:
            self.front = self.rear = 0
        else:
            self.rear = (self.rear + 1) % self.capacity
        self.arr[self.rear] = x
        self.size += 1
        return True
    
    def delete_front(self):
        if self.size == 0:
            return False
        self.front = (self.front + 1) % self.capacity
        self.size -= 1
        return True
    
    def delete_rear(self):
        if self.size == 0:
            return False
        self.rear = (self.rear - 1 + self.capacity) % self.capacity
        self.size -= 1
        return True

File: src/data/courses/dsa/queues/deque/implement-deque/code/java/ImplementDeque.java
class Deque {
    int[] arr;
    int front, rear, size, capacity;
    
    Deque(int k) {
        arr = new int[k];
        capacity = k;
        front = rear = -1;
        size = 0;
    }
    
    boolean insertFront(int x) {
        if (size == capacity) return false;
        if (front == -1) front = rear = 0;
        else front = (front - 1 + capacity) % capacity;
        arr[front] = x;
        size++;
        return true;
    }
    
    boolean insertRear(int x) {
        if (size == capacity) return false;
        if (rear == -1) front = rear = 0;
        else rear = (rear + 1) % capacity;
        arr[rear] = x;
        size++;
        return true;
    }
    
    boolean deleteFront() {
        if (size == 0) return false;
        front = (front + 1) % capacity;
        size--;
        return true;
    }
    
    boolean deleteRear() {
        if (size == 0) return false;
        rear = (rear - 1 + capacity) % capacity;
        size--;
        return true;
    }
}